File: src/10.graph/47.210.ts
export function findOrder(
    numCourses: number,
    prerequisites: number[][]
): number[] {
    // 计算入度
    const inDrgees = new Array(numCourses).fill(0);

    const map = new Map<number, number[]>();

    for (const val of prerequisites) {
        const [end, start] = val;

        inDrgees[end]++;
        if (map.has(start)) {
            map.get(start).push(end);
        } else {
            map.set(start, [end]);
        }
    }

    const queue: number[] = [];

    for (let i = 0; i < inDrgees.length; i++) {
        if (inDrgees[i] === 0) {
            queue.push(i);
        }
    }

    const res: number[] = [];

    while (queue.length) {
        const cur = queue.shift();

        res.push(cur);

        if (!map.has(cur)) {
            continue;
        }

        map.get(cur).forEach((item) => {
            inDrgees[item]--;
            if (inDrgees[item] === 0) {
                queue.push(item);
            }
        });
    }

    return res.length === numCourses ? res : [];
}

File: src/10.graph/47.210.go
package graph

func FindOrder(numCourses int, prerequisites [][]int) []int {
	inDrg := make([]int, numCourses)
	mp := make(map[int][]int)
	for _, v := range prerequisites {
		end, start := v[0], v[1]

		inDrg[end]++

		if _, ok := mp[start]; !ok {
			mp[start] = []int{end}
		} else {
			mp[start] = append(mp[start], end)
		}
	}

	queue := []int{}

	for i, v := range inDrg {
		if v == 0 {
			queue = append(queue, i)
		}
	}

	res := []int{}

	for len(queue) > 0 {
		cur := queue[0]
		queue = queue[1:]
		res = append(res, cur)

		if _, ok := mp[cur]; ok {
			for _, v := range mp[cur] {
				inDrg[v]--

				if inDrg[v] == 0 {
					queue = append(queue, v)
				}
			}
		}
	}

	if len(res) == numCourses {
		return res
	}

	return []int{}
}
